src/libraries/System.Net.Http/src/System/Net/Http/Metrics/HttpMetricsEnrichmentContext.cs

```diff
@@ -0,0 +1,162 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Net.Http.Metrics;
+
+public class HttpMetricsEnrichmentContext
+{
+    private HttpRequestMessage? _request;
+    private HttpResponseMessage? _response;
+    private Exception? _exception;
+    private TagCollection _tags;
+
+    public HttpMetricsEnrichmentContext()
+    {
+        _tags = new TagCollection();
+    }
+
+    public HttpRequestMessage Request
+    {
+        get
+        {
+            EnsureRunningCallback();
+            return _request!;
+        }
+    }
+
+    public HttpResponseMessage? Response
+    {
+        get
+        {
+            EnsureRunningCallback();
+            return _response;
+        }
+    }
+
+    public Exception? Exception
+    {
+        get
+        {
+            EnsureRunningCallback();
+            return _exception;
+        }
+    }
+
+    internal bool InProgress => _request != null;
+
+    public ICollection<KeyValuePair<string, object?>> CustomTags => _tags;
+
+    internal void ApplyEnrichment(HttpRequestMessage request, HttpResponseMessage? response, Exception? exception, ref TagList tags)
+    {
+        if (request._options?.TryGetMetricsEnrichmentCallbackCollection(out List<Action<HttpMetricsEnrichmentContext>>? callbackCollection) != true)
+        {
+            return;
+        }
+
+        _request = request;
+        _response = response;
+        _exception = exception;
+        _tags._tags.Clear();
```

Comment:
If this Clear is because of pooling, we should be clearing before returning the object to the pool rather than when we get the object from the pool.  Two reasons: primarily 1) if the object stores references to other objects, we don't want those to be kept rooted from the pool artificially, and secondarily 2) if the object is newly created, clearing it is a nop.  Is that not possible here for some reason?  Do we not know when we're done with it?